File: src/lib/bot/handlers/daily.ts
// /daily — card of the day. Free, 1/day per user. Updates streak.
// Ported from python-bot/app/handlers/daily.py.

import { db } from '@/lib/db';
import type { TelegramMessage } from '../types';
import { sendMessage } from '../telegram';
import { mainMenuKeyboard } from '../keyboards';
import { t, type Locale } from '../i18n';
import {
  findUserByTelegramId,
  DAILY_CARD_COOLDOWN_HOURS,
} from './_helpers';
import * as crystals from '../services/crystals';
import { drawRandomCards, cardsToJSON } from '../services/tarot';
import { cardOfDay } from '../services/ai';

export async function handleDaily(message: TelegramMessage): Promise<void> {
  if (!message.from) return;
  const tgId = String(message.from.id);
  const user = await findUserByTelegramId(tgId);
  if (!user) {
    await sendMessage(message.chat.id, t('err_unknown_user', 'ru'));
    return;
  }
  const loc = (user.language as Locale) ?? 'ru';
  if (user.isBlocked) {
    await sendMessage(message.chat.id, t('err_blocked', loc));
    return;
  }

  const now = new Date();
  const last = user.lastDailyCardAt;
  if (last && now.getTime() - new Date(last).getTime() < DAILY_CARD_COOLDOWN_HOURS * 3_600_000) {
    const remainingMs = DAILY_CARD_COOLDOWN_HOURS * 3_600_000 - (now.getTime() - new Date(last).getTime());
    const remainingHours = Math.max(0, Math.floor(remainingMs / 3_600_000));
    await sendMessage(
      message.chat.id,
      t('card_of_day_cooldown', loc, { hours: remainingHours }),
      { reply_markup: mainMenuKeyboard(user, loc) },
    );
    return;
  }

  // Update streak + maybe give bonus.
  const bonusInfo = await crystals.checkAndGiveDailyBonus(user.id);

  // Draw a card.
  const cards = drawRandomCards(1, loc);
  const card = cards[0];
  await sendMessage(message.chat.id, t('reading_processing', loc));

  let interpretation: string;
  try {
    interpretation = await cardOfDay({
      name: user.name ?? 'друг',
      zodiac: user.zodiacSign,
      cardName: card.name,
      reversed: card.reversed,
    });
  } catch (e) {
    console.warn('[daily] card_of_day_llm_failed', e);
    const revNote = card.reversed ? ' (перевёрнута)' : '';
    interpretation =
      `🃏 ${card.name}${revNote}\n\n` + t('card_of_day_fallback', loc);
  }

  await db.reading.create({
    data: {
      userId: user.id,
      type: 'card_of_day',
      cards: cardsToJSON(cards),
      interpretation,
      cost: 0,
    },
  });
  await db.user.update({
    where: { id: user.id },
    data: { lastDailyCardAt: now },
  });

  if (bonusInfo.bonus > 0) {
    await sendMessage(
      message.chat.id,
      t('daily_bonus_received', loc, { days: bonusInfo.streak }),
    );
  }

  const revNote = card.reversed ? ' (перевёрнута)' : '';
  await sendMessage(
    message.chat.id,
    `🃏 ${card.name}${revNote}\n\n${interpretation}`,
    { reply_markup: mainMenuKeyboard(user, loc) },
  );
}

File: python-bot/app/handlers/daily.py
"""app.handlers.daily — /daily command (card of the day, free, 1/day, updates streak)."""
from __future__ import annotations

from datetime import datetime, timedelta
from typing import Any, Dict

from aiogram import Router, types
from aiogram.filters import Command

from app.config import settings
from app.db import Queries
from app.i18n import t
from app.keyboards.inline import main_menu_keyboard
from app.services import crystals as crystal_service
from app.services.ai import ai, LLMError
from app.services.tarot import draw_random_cards, cards_to_json
from app.utils.logger import get_logger

log = get_logger("app.handlers.daily")

router = Router(name="daily")


@router.message(Command("daily"))
async def cmd_daily(message: types.Message) -> None:
    """Card of the day — free, once every N hours. Updates streak."""
    if not message.from_user:
        return
    tg_id = str(message.from_user.id)
    user = await Queries.find_user_by_telegram_id(tg_id)
    if not user:
        await message.answer(t("ru", "err_unknown_user"))
        return
    loc = user.get("language", "ru")
    if user.get("isBlocked"):
        await message.answer(t(loc, "err_blocked"))
        return

    now = datetime.utcnow()
    cooldown_hours = settings.daily_card_cooldown_hours

    # Check cooldown.
    last = user.get("lastDailyCardAt")
    if isinstance(last, str):
        try:
            last = datetime.fromisoformat(last.replace("Z", "+00:00"))
        except ValueError:
            last = None
    if isinstance(last, datetime) and last.tzinfo is not None:
        last = last.replace(tzinfo=None)

    if last and (now - last) < timedelta(hours=cooldown_hours):
        remaining_hours = cooldown_hours - int((now - last).total_seconds() // 3600)
        await message.answer(
            t(loc, "card_of_day_cooldown", hours=max(0, remaining_hours)),
            reply_markup=main_menu_keyboard(user),
        )
        return

    # Update streak.
    user = await crystal_service.maybe_daily_bonus(user) or user

    # Draw a card.
    cards = draw_random_cards(1, loc)
    card = cards[0]
    await message.answer(t(loc, "reading_processing"))

    try:
        interpretation = await ai.card_of_day(
            name=user.get("name"),
            zodiac=user.get("zodiacSign"),
            card_name=card.name,
            reversed=card.reversed,
        )
    except LLMError as e:
        log.warning("card_of_day_llm_failed", extra={"err": str(e)})
        # Fallback — short deterministic text.
        reversed_note = " (перевёрнута)" if card.reversed else ""
        interpretation = (
            f"🃏 {card.name}{reversed_note}\n\n"
            + (loc == "en" and "I would sit with this card today. What does it bring up for you? 🌙"
               or "Я бы посидела с этой картой сегодня. Что она в тебе откликает? 🌙")
        )

    # Save reading + update timestamps.
    await Queries.save_reading(
        user_id=user["id"],
        reading_type="card_of_day",
        question=None,
        cards_json=cards_to_json(cards),
        interpretation=interpretation,
        cost=0,
    )
    await Queries.update_user(user["telegramId"], lastDailyCardAt=now)

    # Daily bonus notification (streak >= 3 → +1 already applied in maybe_daily_bonus).
    # Show extra message if a bonus was awarded — we check streakDays.
    if int(user.get("streakDays") or 0) >= 3:
        # The bonus was already applied; just inform the user.
        await message.answer(
            t(loc, "daily_bonus_received", days=user.get("streakDays", 0))
        )

    await message.answer(
        f"🃏 {card.name}{' (перевёрнута)' if card.reversed else ''}\n\n{interpretation}",
        reply_markup=main_menu_keyboard(user),
    )
